localstack/packages/__init__.py

```diff
@@ -0,0 +1,349 @@
+        run(self._debian_cmd_prefix() + ["update"])
+
+    def _debian_install(self, target: InstallTarget) -> None:
+        debian_packages = self._debian_packages()
+        LOG.debug("Downloading packages %s to folder: %s", packages, _DEBIAN_CACHE_DIR)
+        cmd = self._debian_cmd_prefix() + ["-d", "install"] + debian_packages
+        run(cmd)
+        cmd = self._debian_cmd_prefix() + ["install"] + debian_packages
+        run(cmd)
+
+    def _debian_post_process(self, target: InstallTarget) -> None:
+        # TODO maybe remove the debian cache dir here?
+        ...
+
+    def _debian_cmd_prefix(self) -> List[str]:
+        """Return the apt command prefix, configuring the local package cache folders"""
+        return [
+            "apt",
+            f"-o=dir::cache={_DEBIAN_CACHE_DIR}",
+            f"-o=dir::cache::archives={_DEBIAN_CACHE_DIR}",
+            "-y",
+        ]
+
+    def _redhat_get_install_dir(self, target: InstallTarget) -> str:
+        raise SystemNotSupportedException(
+            f"There is no supported installation method for {self.name} on RedHat."
+        )
+
+    def _redhat_packages(self) -> List[str]:
+        raise SystemNotSupportedException(
+            f"There is no supported installation method for {self.name} on RedHat."
+        )
+
+    def _redhat_prepare_install(self, target: InstallTarget) -> None:
+        ...
+
+    def _redhat_install(self, target: InstallTarget) -> None:
+        run(["dnf", "install", "-y"] + self._redhat_packages())
+
+    def _redhat_post_process(self, target: InstallTarget) -> None:
+        run(["dnf", "clean", "all"])
+
+
+class Package(abc.ABC):
+    def __init__(self, name: str, default_version: str):
+        self.name = name
+        self.default_version = default_version
+
+    def get_executables_path(self, version: str | None = None) -> str | None:
+        return self.get_installer(version).get_executables_path()
+
+    def get_installed_dir(self, version: str | None = None) -> str | None:
+        return self.get_installer(version).get_installed_dir()
+
+    def install(self, version: str | None = None, target: Optional[InstallTarget] = None) -> None:
+        self.get_installer(version).install(target)
+
+    def get_installer(self, version: str | None = None) -> PackageInstaller:
+        if not version:
+            version = self.default_version
+        if version not in self.get_versions():
+            raise NoSuchVersionException()
+        return self._get_installer(version)
+
+    def get_versions(self) -> List[str]:
+        raise NotImplementedError()
+
+    def _get_installer(self, version: str) -> PackageInstaller:
+        raise NotImplementedError()
+
+    def __str__(self):
+        return self.name
+
+
+class PackageRepository(PluginManager):
+    def __init__(self):
+        super().__init__(namespace=PLUGIN_NAMESPACE)
+
+
+class PackagesPlugin(Plugin):
+    service: str
```

Comment:
is this the service the package should map to, or is this just the name of the package it installs? in other words: are you intending a 1:1 mapping between packages and services?

Reply:
It is the _service_ this packages plugin refers to. It is a 1:n mapping between service and packages (as you can see with the return type in the `get_packages` signature).
An example usage of the `PackagesPlugin` can already be seen in `opensearch/plugins.py`.

Reply:
Ok! Will be interesting to see how this develops :)